quinn-proto/src/assembler.rs

```diff
@@ -86,12 +86,18 @@ impl Assembler {
         let new = BinaryHeap::with_capacity(self.data.len());
         let old = mem::replace(&mut self.data, new);
         for chunk in old.into_sorted_vec().into_iter().rev() {
-            if offset + (buffer.len() as u64) < chunk.offset {
+            let end = offset + (buffer.len() as u64);
+            let overlap = if let Some(overlap) = end.checked_sub(chunk.offset) {
+                overlap
+            } else {
                 let bytes = buffer.split().freeze();
                 self.data.push(Chunk { offset, bytes });
                 offset = chunk.offset;
+                0
+            };
+            if let Some(bytes) = chunk.bytes.get(overlap as usize..) {
+                buffer.extend_from_slice(bytes);
             }
```

Comment:
Perhaps this would be cleaner if `extend_from_slice` was moved into both branches above? Only the first instance would need the checked `get` call, and it'd save the weird dance to get an overlap value out.

Reply:
Good idea, updated.